K-evolution/auxiliary_library.py

```diff
@@ -835,70 +845,54 @@ def recursive_basis(depth, seed_op, Hamiltonian, rho0):
-    gram_matrix = [[sc_prod(op2, op1, rho0) for op2 in basis] for op1 in basis]
+
+    dim = len(basis)
+    
     hermitian_basis = [linalg.norm(op1 - op1.dag()) < 1e-10 for op1 in basis]
-    mean0_centered_ops = [np.real((rho0 * op1).tr()-0) > 1e-10 for op1 in basis]
-    
-    for i in range(len(basis)): 
-        mu0i= mean0_centered_ops[i]
-        if (-1e-5 > np.real(mu0i) > 1e-5 and abs(np.imag(mu0i)) > 1e-10):
-            print("Not mean-normalized operator at", i, "-th level")
-            print((rho0 * basis[i]).tr())
-        if (abs(gram_matrix[i][i] - 1) < 10**-10):
-            all_gram_diagonals_are_one = True
-        else:
-            all_gram_diagonals_are_one = False
-            print("The", i,"-th operator is not normalized \n")
+    assert np.all(hermitian_basis), ("Not all the operators are "
+                                     f"hermitician:\n {hermitian_basis}")
+
+    null_averages = [np.real((rho0 * op1).tr()) < 1e-10 for op1 in basis]
+    assert all(null_averages[1:]), ("Some operators do not have a null average:\n" 
+                                    f"{null_averages}")
     
-    if (linalg.norm((np.identity(len(basis)) - gram_matrix) < 10**-10)):
-        all_ops_orth = True
-    else:
-        all_ops_orth = False
-        print("Not all operators are pair-wise orthogonal")
-        
-    if (np.all(hermitian_basis) and all_gram_diagonals_are_one and all_ops_orth):
-        print("The basis is orthonormal and hermitian")
-    return basis, qutip.Qobj(gram_matrix)
+    gram_matrix = [[sc_prod(op2, op1, rho0) for op2 in basis] for op1 in basis]
+    
+    normalized = [abs(gram_matrix[i][i]-1.)<1e-10  for i in range(dim)]
+    assert all(normalized), ("Some operators in the basis are not normalized:\n"
+                             f"{normalized}")
+
+    assert (linalg.norm(
+        (np.identity(dim) - gram_matrix) < 10**-10)), "Not all operators are pair-wise orthogonal"
+    
+    print("The basis is orthonormal and hermitian")
+    return True
 
 # In [18]:
 
-def build_rho0_from_basis(basis):
-    phi0 = [0] + [np.random.rand() for i in range(len(basis)-1)]
-    rho0 = (-sum( f*op for f,op in zip(phi0, basis))).expm()
-    phi0[0] = np.log(rho0.tr())
-    rho0 = (-sum( f*op for f,op in zip(phi0, basis))).expm()
-    
-    if (is_density_op(rho0)):
-        pass
-    else:
-        if not ev_checks(rho0):
-            print("rho0: not positive defined")
-        if (non_hermitianess_measure(rho0) < 1e-6):
-            rho0 = .5 * (rho0 + rho0.dag())
-        if (rho0.tr() != 1):
-            rho0 = rho0/rho0.tr()            
+def build_rho0_from_basis(basis, temp=1.):
+    phi0 = [np.random.rand()/temp for b in basis]
+    k0 = -sum( f*op for f,op in zip(phi0, basis))
+    k0 = k0 - max(k0.eigenenergies())
+    rho0 = (k0).expm()
+    normalizacion = rho0.tr()
+    rho0 = rho0 / normalizacion
+    assert is_density_op(rho0, verbose=True), "rho is not a density matrix."
```

Comment:
Let's be less tolerant. If the initial state is not a density operator, let's abort the evaluation.